include/jemalloc/internal/peak.h

```diff
@@ -0,0 +1,32 @@
+#ifndef JEMALLOC_INTERNAL_PEAK_H
+#define JEMALLOC_INTERNAL_PEAK_H
+
+typedef struct peak_s peak_t;
+struct peak_s {
+	uint64_t cur_max;
+	uint64_t adjustment;
+};
+
+#define PEAK_INITIALIZER {0, 0}
+
+static inline uint64_t
+peak_max(peak_t *peak) {
+	return peak->cur_max;
+}
+
+static inline void
+peak_update(peak_t *peak, uint64_t alloc, uint64_t dalloc) {
+	int64_t candidate_max = (int64_t)(alloc - dalloc - peak->adjustment);
+	if (candidate_max > (int64_t)peak->cur_max) {
+		peak->cur_max = candidate_max;
+	}
```

Comment:
Casting to `int64_t` might risk underflow / overflow issue. Can the logic here be implemented differently?

Reply:
This is actually pretty safe -- the only underflow/overflow issues occur when the net bytes allocated by a thread exceed INT64_MAX between reset calls. (I.e. we're resilient to thread.allocated/deallocated wraparound outside of that case).